File: lib/domain/field/ports/field_list_output_port.rb
# frozen_string_literal: true

module Domain
  module Field
    module Ports
      class FieldListOutputPort
        def on_success(fields)
          raise NotImplementedError, "Subclasses must implement on_success"
        end

        def on_failure(error_dto)
          raise NotImplementedError, "Subclasses must implement on_failure"
        end
      end
    end
  end
end

File: crates/agrr-domain/src/field/ports/field_list_output_port.rs
use crate::field::results::FarmFieldsList;
use crate::shared::dtos::error::Error;
use crate::shared::policies::policy_permission_denied::PolicyPermissionDenied;

/// Ruby: `Domain::Field::Ports::FieldListOutputPort`
pub trait FieldListOutputPort {
    fn on_success(&mut self, result: FarmFieldsList);
    fn on_failure(&mut self, error: ListFailure);
}

#[derive(Debug, Clone, PartialEq)]
pub enum ListFailure {
    Policy(PolicyPermissionDenied),
    Error(Error),
}
